server/src/main/java/org/elasticsearch/search/query/QueryPhase.java

```diff
@@ -341,5 +559,77 @@ private static boolean canEarlyTerminate(IndexReader reader, SortAndFormats sort
         return true;
     }
 
+    /**
+     * Returns true if more than 50% of data in the index have the same value
+     * The evaluation is approximation based on finding the median value and estimating its count
+     * Returns true if the total count of median values is greater or equal to half of the total count of documents
+     */
+    static boolean indexFieldHasDuplicateData(IndexReader reader, String field) throws IOException {
+        long globalDocCount = 0;
+        long globalMedianCount = 0;
+        for (LeafReaderContext lrc : reader.leaves()) {
+            PointValues pointValues = lrc.reader().getPointValues(field);
+            if (pointValues == null) continue;
+            int docCount = pointValues.getDocCount();
+            if (docCount <= 512) { // skipping small segments as estimateMedianCount doesn't work well on them
+                continue;
+            }
+            assert(pointValues.size() == docCount); // TODO: modify the code to handle multiple values
+            globalDocCount += docCount;
+            long medianValue = estimateMedianValue(pointValues);
+            long medianCount = estimatePointCount(pointValues, medianValue, medianValue);
+            globalMedianCount += medianCount;
+        }
+        return (globalMedianCount >= globalDocCount/2);
+    }
+
+    static long estimateMedianValue(PointValues pointValues) throws IOException {
+        long minValue = LongPoint.decodeDimension(pointValues.getMinPackedValue(), 0);
+        long maxValue = LongPoint.decodeDimension(pointValues.getMaxPackedValue(), 0);
+        while (minValue < maxValue) {
+            long avgValue = Math.floorDiv(minValue, 2) + Math.floorDiv(maxValue, 2); // to avoid overflow first divide each value by 2
+            long countLeft = estimatePointCount(pointValues, minValue, avgValue);
+            long countRight = estimatePointCount(pointValues, avgValue + 1, maxValue);
```

Comment:
I wonder that this method might be quite expensive by calling estimatePointCount in a loop, can we break the for loop as soon as countLeft + countRight <= globalDocCount/2 (which might require to rename this method to something like `boolean hasValueWithCountGreaterThan(PointValues values, long threshold)` or something along those lines).

Reply:
Thanks. I was thinking the same, and implemented it in 5870fd7 , but @jimczi thinks that this is incorrect logic.  Now thinking again, I think that may be it is correct, and I should have the commit 5870fd7  back.

Reply:
I don't think the logic is the same in https://github.com/elastic/elasticsearch/commit/5870fd7a48d65c5daa4c995db97d0a44e069961e. The idea that Adrien mentioned is to early terminate if countLeft+countRight is greater than the global doc count divided by 2 while your commit was checking the local doc count and only on one side (right or left). I also wonder if we need to check that `countLeft > 0 || countRight > 0` before checking the early termination ?